### ISSUE TITLE
Add support for other methods of authentication

### ISSUE THREAD

**Issue by DanWebb** [DanWebb/uphold-sdk-node#1]
Currently uphold-sdk-node only supports the [web application flow](https://uphold.com/en/developer/api/documentation/#web-application-flow) for authenticating, it should also support [Personal Access Tokens (PAT)](https://uphold.com/en/developer/api/documentation/#personal-access-tokens-pat) and [Basic Authentication](https://uphold.com/en/developer/api/documentation/#basic-authentication)

**Comment by andreipopovici** [comment]
Hey there,

I tried a simple implementation: https://github.com/andreipopovici/uphold-sdk-node/commit/0e6755e58e11bdfa9a8d9485e481dd3704a0b587

However, it would be easiest to test this if you could add your own PAT and credentials, to match the key & secret already used.

What approach would you suggest for 2FA authentication? Since neither PAT nor BasicAuth are recommended for non-sandbox usage, would it be best to simply ignore 2FA and update the documentation to explain this? Users could go and disable 2FA in their sandbox account before using the SDK.

Also, since it's a Node SDK, would you be open to adding some ES6 (say, Object.assign() in the tests)?

**Comment by DanWebb** [comment]
Hi @andreipopovici, after talking with Jorge from Uphold I've decided it'd only be worth implementing the PAT authentication as an alternative method for use with private scripts as use of basic auth isn't recommended and will eventually be depreciated. PAT authentication could be used outside of the sandbox environment just probably not applicable on a public facing application so I've chosen to support 2FA also.

Thanks a lot for taking the time to share your implementation, I've started adding the methods for PAT authentication but am currently having trouble finding the correct api route for it as [the ones documented](https://uphold.com/en/developer/api/documentation/#creating-a-pat) are giving me a 404 not found error, don't know if you can be of any help there? I created a support ticket anyway.

For now the sdk is written without using ES6 specific features but I wouldn't be against adding them.

**Comment by DanWebb** [comment]
No worries got a successful response, I was testing with postman but seems to work when attempting the request outside of postman.

**Comment by andreipopovici** [comment]
I'm still getting a 404 with Postman, weird that it works outside of that. I wonder what Postman does to send an invalid call... huh.

Thanks for adding PAT-related calls to the SDK. However, there might be a better way to handle 2FA in the `createPAT` callback, seeing as uphold returns `X-Bitreserve-OTP: Required` if they've just sent a code. Unless they plan to switch to TOTP (for use with Google Authenticator or Authy) instead of sending SMS, I'm not sure how one could programmatically (or manually) request a code, other than attempting to run the script, input the code, then run it again.

Out of coder's curiosity, why did you use a header with `Authorization: Basic` instead of using the `auth` property provided by the `request` library?

**Comment by DanWebb** [comment]
You're right with the X-Bitreserve-OTP header, I was aware of this but when implementing it last night I thought just checking for the Unauthorized error would be enough. After checking into it I've found that "Unauthorized" is also returned when an incorrect username/password combination is entered so it won't do. I've updated the sdk to handle the X-Bitreserve-OTP response header here: 325fe92

Thanks for the tip with the auth property, I had never come across it before and have updated the sdk to use it as it's less verbose.
